graphene_django/static/graphene_django/graphiql.js

```diff
@@ -39,27 +54,31 @@
 
   // Defines a GraphQL fetcher using the fetch API.
   function graphQLFetcher(graphQLParams) {
-    var headers = {
-      'Accept': 'application/json',
-      'Content-Type': 'application/json'
-    };
-    if (csrftoken) {
-      headers['X-CSRFToken'] = csrftoken;
+    function getFetch(headers) {
+      return fetch(fetchURL, {
+        method: 'post',
+        headers: headers,
+        body: JSON.stringify(graphQLParams),
+        credentials: 'include'
+      });
     }
-    return fetch(fetchURL, {
-      method: 'post',
-      headers: headers,
-      body: JSON.stringify(graphQLParams),
-      credentials: 'include',
-    }).then(function (response) {
-      return response.text();
-    }).then(function (responseBody) {
-      try {
-        return JSON.parse(responseBody);
-      } catch (error) {
-        return responseBody;
-      }
-    });
+    return getFetch(headers)
+      .then(function(response) {
+        console.log(headers);
+        return response.text();
+      })
+      .then(function(responseBody) {
+        try {
+          return JSON.parse(responseBody);
+        } catch (error) {
+          if (cookies.csrftoken && cookies.csrftoken.length) {
+            headers['X-CSRFToken'] = cookies.csrftoken.pop();
+            console.log('retry', headers)
+            return getFetch(headers);
+          }
```

Comment:
There could be other reasons why the GraphQL endpoint doesn't return valid JSON so always assuming it's to do with a csrf token issue doesn't feel right.

Reply:
see: https://stackoverflow.com/questions/576535/cookie-path-and-its-accessibility-to-subfolder-pages

From what I understand, the cookie sub path `/2/m/2` would have both it's subpath cookie and the root path cookie. `/`.

The two paths have different database multitenancy requirements. It would probably be more accurate to read the cookie path directly, instead of just reading, checking them all?

Reply:
Yes I understand that cookies can have different paths but looking at the Django source code the csrf cookie should be set to a path that is defined in the settings: https://github.com/django/django/blob/5a68f024987e6d16c2626a31bf653a2edddea579/django/middleware/csrf.py#L191

So I don't think it's possible to have 2 csrf cookies at the same time and I'm wondering how that came about?

Reply:
endpoints per multi tenancy. `localhost:8000/:multitenantid/endpoints/here` so multiple users can get different csrftoken cookies on a single machine ... it's not the default of django, but it is a real use case.

Reply:
Ah I didn't realise that was possible! Could you not host the graphql endpoint per tenant as well?